src/webui/api/torrentscontroller.cpp

```diff
@@ -1475,6 +1475,24 @@ void TorrentsController::addTagsAction()
     }
 }
 
+void TorrentsController::setTagsAction()
+{
+    requireParams({u"hashes"_s, u"tags"_s});
+
+    const QStringList hashes {params()[u"hashes"_s].split(u'|', Qt::SkipEmptyParts)};
+    const QStringList tags {params()[u"tags"_s].split(u',', Qt::SkipEmptyParts)};
+
+    const TagSet newTags {tags.begin(), tags.end()};
+    applyToTorrents(hashes, [&newTags](BitTorrent::Torrent *const torrent)
+    {
+        for (const Tag &tag : asConst(torrent->tags()))
+            if (!newTags.contains(tag))
+                torrent->removeTag(tag);
+        for (const Tag &tag : newTags)
+            torrent->addTag(tag);
+    });
```

Comment:
Excuse me, I missed a pair of braces in my previous suggestion.
Also, optimize it a bit:
```suggestion
    TagSet newTags {tags.begin(), tags.end()};
    applyToTorrents(hashes, [&newTags](BitTorrent::Torrent *const torrent)
    {
        for (const Tag &tag : asConst(torrent->tags()))
        {
            if (newTags.erase(tag) == 0)
                torrent->removeTag(tag);
        }
        for (const Tag &tag : newTags)
            torrent->addTag(tag);
    });
```

Reply:
That's a clever optimization! Thanks for being patient and so helpful reviewing!